app/controllers/test/saml_test_controller.rb

```diff
@@ -0,0 +1,147 @@
+      end
+
+      render template: 'test/saml_test/decode_response.html.slim',
+             locals: { is_valid: is_valid, response: response }
+    end
+    # rubocop:enable AbcSize, MethodLength
+
+    # Method to handle IdP initiated logouts
+    # rubocop:disable AbcSize, MethodLength
+    def decode_slo_request
+      if params[:SAMLRequest]
+        logout_request = OneLogin::RubySaml::SloLogoutrequest.new(params[:SAMLRequest])
+
+        doc = Saml::XML::Document.parse(logout_request.document.to_s)
+        is_valid = doc.valid_signature?(Rails.application.secrets.saml_cert)
+
+        if is_valid
+          logger.info "IdP initiated Logout for #{logout_request.name_id}"
+          # Generate a response to the IdP.
+          logout_request_id = logout_request.id
+          logout_response = OneLogin::RubySaml::SloLogoutresponse.new
+          saml_response = logout_response.create(
+            saml_settings,
+            logout_request_id,
+            nil,
+            RelayState: params[:RelayState]
+          )
+          redirect_to saml_response and return
+        else
+          response = doc.errors.to_s
+          render template: 'test/saml_test/decode_response.html.slim',
+                 locals: { is_valid: is_valid, response: response }
+        end
+      elsif params[:SAMLResponse]
+        decode_response
+      end
+    end
+    # rubocop:enable AbcSize, MethodLength
+
+    private
+
+    # rubocop:disable AbcSize, MethodLength
+    def saml_settings
+      settings = OneLogin::RubySaml::Settings.new
+
+      # SP settings
+      settings.assertion_consumer_service_url = 'http://localhost:3000/test/saml/decode_assertion'
+      settings.assertion_consumer_logout_service_url = 'http://localhost:3000/test/saml/decode_slo_request'
+      settings.certificate = Rails.application.secrets.saml_cert
+      settings.private_key = private_key
+      settings.authn_context = Saml::Idp::Constants::LOA1_AUTHNCONTEXT_CLASSREF
+
+      # SP + IdP Settings
+      settings.issuer = 'https://upaya-dev.ngrok.io'
+      settings.security[:logout_requests_signed] = true
+      settings.security[:embed_sign] = true
+      settings.security[:digest_method] = 'http://www.w3.org/2001/04/xmlenc#sha256'
+      settings.security[:signature_method] = 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'
+      settings.name_identifier_format = 'urn:oasis:names:tc:SAML:2.0:nameid-format:persistent'
+      settings.double_quote_xml_attribute_values = true
+      # IdP setting
+      settings.idp_sso_target_url = 'https://upaya-dev.ngrok.io/api/saml/auth'
+      settings.idp_slo_target_url = 'https://upaya-dev.ngrok.io/api/saml/logout'
+      settings.idp_cert_fingerprint = fingerprint
+
+      settings
+    end
+    # rubocop:enable AbcSize, MethodLength
+
+    def private_key
+      OpenSSL::PKey::RSA.new(
+        File.read(Rails.root + 'config/saml.key.enc'),
+        Rails.application.secrets.saml_passphrase
+      ).to_pem
+    end
```

Comment:
nitpick: it seems like we should make this a constant or define it once and reuse it elsewhere:

* https://github.com/18F/identity-idp/blob/master/config/initializers/omniauth.rb#L12
* https://github.com/18F/identity-idp/pull/8/files#diff-09343f25b4506017cf86bd0730b6a752R8
* https://github.com/18F/identity-idp/pull/8/files#diff-61a7fc82d9813199a02d76a2c0f87860R31